symfit/core/argument.py

```diff
@@ -111,3 +113,60 @@ class Variable(Argument):
     # Variable index to be assigned to generated nameless variables
     _argument_name = 'var'
     __slots__ = ()
+
+class IndexedArgument(Indexed):
+    def __new__(cls, base, *args, **kw_args):
+        """
+        __new__ has to be overwritten to be duck typing compliant: just try it!
+        """
+        return Expr.__new__(cls, base, *args, **kw_args)
+
+    @property
+    def free_symbols(self):
+        """
+        In sympy, Indexed objects return their base and indices as free symbols.
+        We would like the Indexed object to be seen as fundamental however.
+        :return:
+        """
+        return {self}
+
+
+class IndexedVariable(IndexedArgument):
+    pass
+
+
+class IndexedParameter(IndexedArgument):
+    pass
+
+
+class IndexedArgumentBase(IndexedBase):
+    __slots__ = ()
+    _indexed_class = IndexedArgument
+
+    def __getitem__(self, indices, **kw_args):
+        try:
+            # Use this to do all sanity checking on the arguments, but if they
+            # go through, we build a self._indexed_class instead.
+            super(IndexedArgumentBase, self).__getitem__(indices, **kw_args)
+        except IndexException as err:
+            raise err
+        else:
+            if is_sequence(indices):
+                return self._indexed_class(self, *indices, **kw_args)
+            else:
+                return self._indexed_class(self, indices, **kw_args)
+
+    @property
+    def free_symbols(self):
+        """
+        In sympy, IndexedBase objects return their label as free symbols.
+        We would like the Indexed object to be seen as fundamental however.
+        :return:
+        """
+        return {self}
+
+class IndexedParameterBase(IndexedArgumentBase):
+    _indexed_class = IndexedParameter
+
+class IndexedVariableBase(IndexedArgumentBase):
+    _indexed_class = IndexedVariable
```

Comment:
Yeah they are used, when you want to create an Indexed object in sympy you need to do the following:
```python
a = IndexedBase('a')
i = Idx('i')
type(a[i]) -> Indexed
type(a) -> IndexedBase
```
I follow the same structure with these objects, except we again need variables and parameters to be identifiable, hence these new objects. Technically speaking I think we could do without `IndexedArgumentBase` and `IndexedArgument`, just like we could do without `Argument`, those are just there in case we ever need to discriminate ordinary `Symbol` from `symfit` ones.

Reply:
Ah, that starts to make sense. I'd call `i` an `Index` instead of `Indexed`, but meh. It's better to stick to their naming scheme.
Leave Arguments in, we'll need them later.

Reply:
Be careful, `i` is an `Idx`, not an `Indexed`. `Indexed` is what is returned by indexing an `IndexedBase`. 

I would've also just called `Idx` `Index`, but indeed it's better to stick to the same structure.

Reply:
Right, that makes more sense. Now for the $1m question: Why is it an Indexed, and not an e.g. Variable?